src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Media/Imaging/WeakReferenceCache.cs

```diff
@@ -0,0 +1,84 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+
+using System.Threading;
+
+namespace System.Windows.Media.Imaging;
+
+/// <summary>
+/// Provides keyed-caching for different objects stored as weak references.
+/// </summary>
+/// <remarks>The cache operations are thread-safe.</remarks>
+internal sealed class WeakReferenceCache<K, V> where K : notnull
+                                               where V : class
+{
+    /// <summary>
+    /// Max number of elements stored in the cache.
+    /// </summary>
+    private const int MaxCacheSize = 300;
+
+    private readonly Dictionary<K, WeakReference<V>> _cacheStore = new();
+    private readonly Lock _cacheLock = new();
+
+    /// <summary>
+    /// Adds an object to cache
+    /// </summary>
+    public void AddToCache(K key, WeakReference<V> value)
```

Comment:
I feel like `ToCache` is redundant here:
```suggestion
    public void Add(K key, WeakReference<V> value)
```

Reply:
Makes sense